benchmarks/rust/src/bin/kv_indexer.rs

```diff
@@ -0,0 +1,254 @@
+    }
+    Ok(value)
+}
+
+#[derive(Parser, Debug)]
+#[clap(version, about, long_about = None)]
+struct Args {
+    /// Path to mooncake trace.
+    mooncake_trace_path: String,
+
+    /// Amount of shards to use for the indexer. This directly corresponds to the degree of parallelism.
+    /// More parallelism means higher event processing rates at the expense of slower request rates.
+    #[arg(short, long, default_value_t = 1)]
+    num_shards: usize,
+
+    /// A mock for remove block events.
+    /// When a stored event is created, we also create a corresponding remove event that removes a fraction of the blocks that were stored.
+    // TODO: This is a bit of a hack to simulate block removal. We should instead more closely model an LRU behavior.
+    #[arg(short, long, default_value_t = 0.25, value_parser = parse_percent)]
+    removal_fraction: f64,
+}
+
+fn load_trace(args: &Args) -> (Vec<KvCacheEvent>, Vec<Vec<u64>>) {
+    println!("Loading trace from {}", args.mooncake_trace_path);
+
+    let lines: Vec<String> = read_to_string(&args.mooncake_trace_path)
+        .unwrap()
+        .lines()
+        .map(String::from)
+        .collect();
+
+    println!("Reading {} events", lines.len());
+
+    let mut events = Vec::new();
+    let mut sequences = Vec::new();
+
+    // Track the blocks that are currently in our simulated cache.
+    let mut seen_blocks = HashSet::new();
+
+    let mut event_id = 0;
+
+    for line in lines {
+        let json_value: serde_json::Value = serde_json::from_str(&line).unwrap();
+
+        // Extract hash ids from mooncake.
+        let hash_ids = json_value["hash_ids"]
+            .as_array()
+            .unwrap()
+            .iter()
+            .map(|x| x.as_u64().unwrap())
+            .collect::<Vec<_>>();
+
+        sequences.push(hash_ids.clone());
+
+        let block_hashes;
+        let parent_hash;
+
+        // If there are new blocks (that aren't already in our simulated cache), create a store blocks event.
+        if let Some(start) = hash_ids.iter().position(|x| !seen_blocks.contains(x)) {
+            block_hashes = hash_ids[start..].to_vec();
+            if start == 0 {
+                parent_hash = None;
+            } else {
+                parent_hash = Some(ExternalSequenceBlockHash(hash_ids[start - 1]));
+            }
+        } else {
+            continue;
+        };
+
+        // Build our stored blocks event.
+        let event_data = KvCacheEventData::Stored(KvCacheStoreData {
+            parent_hash,
+            blocks: block_hashes
+                .iter()
+                .map(|x| KvCacheStoredBlockData {
+                    block_hash: ExternalSequenceBlockHash(*x),
+                    tokens_hash: LocalBlockHash(*x),
+                })
+                .collect(),
+        });
+
+        events.push(KvCacheEvent {
+            event_id,
+            data: event_data,
+        });
+
+        event_id += 1;
+
+        // Now, we want to simulate block removal.
```

Comment:
we could also use the kvbm here, use logical blocks with no workers which could also in theory simulate offload times.

add to this the router bits. 

we could essentially host virtual nodes with multipe instances of kvbms, have them hooked up to the index over direct channels instead of nats.

you could in theory run a load-gen simulation thru the indexer --> kvbm --> events --> indexer to complete the loop.